ir/src/constraints/graph.rs

```diff
@@ -46,23 +46,64 @@ impl AlgebraicGraph {
         }
     }
 
+    /// TODO: docs
+    pub fn node_details(
+        &self,
+        index: &NodeIndex,
+        default_domain: ConstraintDomain,
+    ) -> Result<(TraceSegment, ConstraintDomain), SemanticError> {
+        // recursively walk the subgraph and infer the trace segment and domain
+        match self.node(index).op() {
+            Operation::Constant(_) => Ok((DEFAULT_SEGMENT, default_domain)),
+            Operation::PeriodicColumn(_, _) => {
+                if !(default_domain.is_integrity()) {
```

Comment:
nit: wouldn't a more readable check be `default_domain.is_boundary()` (instead of `!default_domain.is_integrity()`)?

Or if we keep `!default_domain.is_integrity()`, I'd remove the extra parenthesis around `default_domain.is_integrity()`.